### PR TITLE
PromotionRun should create a SnapshotEnvironmentBinding (for app/env) if it doesn't already exist

### PR DESCRIPTION
**Pull request by jopit**
#### Description:

- Updated the PromotionRun reconciliation to always create a SnapshotEnvironmentBinding (SEB) for the target application/environment, if that SEB doesn't already exist.
- Created unit tests for the new function
- Created e2e tests for the new function

#### Link to JIRA Story (if applicable):

https://issues.redhat.com/browse/GITOPSRVCE-239

### CODE REVIEW

File: appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ import (
 	"time"
 
 	"github.com/go-logr/logr"
+	applicationv1alpha1 "github.com/redhat-appstudio/application-service/api/v1alpha1"
```
Comment (drpaneas) [inline appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go:26]: https://github.com/redhat-appstudio/managed-gitops/pull/261 is replacing application-service repository with appplication-api. You will most likely need to change this.

File: tests-e2e/fixture/fixture.go [hunk 2]
```diff
@@ -11,6 +11,7 @@ import (
 	"time"
 
 	argocdoperator "github.com/argoproj-labs/argocd-operator/api/v1alpha1"
+	applicationv1alpha1 "github.com/redhat-appstudio/application-service/api/v1alpha1"
```
Comment (drpaneas) [inline tests-e2e/fixture/fixture.go:14]: Same comment here.

**Review by drpaneas — COMMENTED** [review]
logic and tests looks fine. My only comment is related to the import that we are replacing in this sprint.

File: appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go [hunk 3]
```diff
@@ -412,10 +420,48 @@ func locateTargetManualBinding(ctx context.Context, promotionRun appstudioshared
...
+	err := k8sClient.Create(ctx, &binding)
+	if err != nil {
+		return appstudioshared.SnapshotEnvironmentBinding{}, err
+	}
+
+	return binding, nil
+}
 
+func createBindingName(promotionRun *appstudioshared.PromotionRun) string {
+	return strings.ToLower(promotionRun.Spec.Application + "-" + promotionRun.Spec.ManualPromotion.TargetEnvironment + "-generated-binding")
```
Comment (jgwest) [inline appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go:464]: It looks like this function will potentially have an issue with long names (names greater that >250 chars). BUT, it wasn't part of the story criteria, so I'm happy to either: a) open a new story for it for next sprint or b) handle it as part of this story/PR. It's up to you. Any preference?

This is what we did for a similar issue on https://issues.redhat.com/browse/GITOPSRVCE-183

File: tests-e2e/fixture/fixture.go [hunk 4]
```diff
@@ -601,6 +601,11 @@ func GetKubeClient(config *rest.Config) (client.Client, error) {
 		return nil, err
 	}
 
+	err = appstudiosharedv1.AddToScheme(scheme)
```
Comment (jgwest) [inline tests-e2e/fixture/fixture.go:604]: Same here, think we've already got this covered :detective:

**Review by jgwest — CHANGES_REQUESTED** [review]
Nice work, thanks @jopit! Some minor tweaks requested then good to go. 

See my one comment about how we want to handle the 'generate name' function that might overflow the maximum K8s resource name, and whether you want to handle it as part of this story or a new one. Up to you, just let me know.

File: appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go [hunk 5]
```diff
@@ -412,10 +420,48 @@ func locateTargetManualBinding(ctx context.Context, promotionRun appstudioshared
...
+			},
+		},
+		Spec: appstudioshared.SnapshotEnvironmentBindingSpec{
+			Application: promotionRun.Spec.Application,
+			Environment: promotionRun.Spec.ManualPromotion.TargetEnvironment,
+			Snapshot:    promotionRun.Spec.Snapshot,
+			Components:  components,
+		},
+	}
+	err := k8sClient.Create(ctx, &binding)
```
Comment (jgwest) [inline appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go:455]: There are a couple items we can log here:
-  We usually call `LogAPIResourceChangeEvent` whenever we are creating/modifying/deleting a resource, we can dd a call here for that. 
    - These log items are mostly for auditing purposes (allow us to keep a record of changes made to the cluster for security purposes.
- It couldn't hurt to add an INFO-level log `log.Info('Created SnapshotEnvironmentBinding", "application", "(app id), "environment", "environment id")`
    - These log statements are more dev/SRE focus, helping us debug the code.

File: appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go [hunk 6]
```diff
@@ -412,10 +420,48 @@ func locateTargetManualBinding(ctx context.Context, promotionRun appstudioshared
 		}
 	}
 
-	return appstudioshared.SnapshotEnvironmentBinding{},
-		fmt.Errorf("unable to locate binding with application '%s' and target environment '%s'",
-			promotionRun.Spec.Application, promotionRun.Spec.ManualPromotion.TargetEnvironment)
+	// Binding not found, so create it
+	components := []appstudioshared.BindingComponent{}
+	componentList := appstudioshared.ComponentList{}
+	if err := k8sClient.List(ctx, &componentList); err != nil {
```
[the remark this answers is not in the document]
Comment (jopit) [reply appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go]: Fixed

File: appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller_test.go [hunk 7]
```diff
@@ -37,6 +40,8 @@ var _ = Describe("SnapshotEnvironmentBinding Reconciler Tests", func() {
 				err := tests.GenericTestSetup()
 			Expect(err).To(BeNil())
 
+			err = appstudiosharedv1.AddToScheme(scheme)
```
[the remark this answers is not in the document]
Comment (jopit) [reply appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller_test.go]: Fixed.  Duplicate was a consequence of rebasing on top of the change which moved the api into a separate repo.

File: appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go (hunk 3 above)
Comment (jopit) [reply appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go]: I've updated the function to ensure the name is never greater than 250 chars.

File: tests-e2e/fixture/fixture.go (hunk 4 above)
Comment (jopit) [reply tests-e2e/fixture/fixture.go]: Fixed.  Duplicate was a consequence of rebasing on top of the change which moved the api into a separate repo.

File: appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go (hunk 5 above)
Comment (jopit) [reply appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go:457]: Added logging for these items

File: appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go [hunk 8]
```diff
@@ -396,12 +406,12 @@ func checkForExistingActivePromotions(ctx context.Context, reconciledPromotionRu
 	return nil
 }
 
-func locateTargetManualBinding(ctx context.Context, promotionRun appstudioshared.PromotionRun, k8sClient client.Client) (appstudioshared.SnapshotEnvironmentBinding, error) {
+func locateOrCreateTargetManualBinding(ctx context.Context, promotionRun appstudioshared.PromotionRun, k8sClient client.Client, log logr.Logger) (appstudioshared.SnapshotEnvironmentBinding, error) {
```
Comment (drpaneas) [inline appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go:409]: This file imports:

```go
"github.com/go-logr/logr"
"sigs.k8s.io/controller-runtime/pkg/log"
```

There is overshadowing of the `log` package with the `log` variable. And this is happening in a lot of places in our code, so at some point we have to address it.

**Review by drpaneas — CHANGES_REQUESTED** [review]
one final small request: Fix the `log` overshadowing the pkg with the variable.

File: appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go (hunk 8 above)
Comment (jopit) [reply appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go]: Fixed.

**Review by drpaneas — APPROVED** [review]
/lgtm

**Review by jgwest — CHANGES_REQUESTED** [review]
Looks great, one final change then good to go.

File: appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go [hunk 9]
```diff
@@ -412,10 +422,59 @@ func locateTargetManualBinding(ctx context.Context, promotionRun appstudioshared
...
+		"environment", promotionRun.Spec.ManualPromotion.TargetEnvironment)
+
+	return binding, nil
+}
 
+func createBindingName(promotionRun *appstudioshared.PromotionRun) string {
+	name := strings.ToLower(promotionRun.Spec.Application + "-" + promotionRun.Spec.ManualPromotion.TargetEnvironment + "-generated-binding")
+	if len(name) > 250 {
+		// 'suffix' will have a length of 32
+		hash := md5.Sum([]byte(name)) //#nosec G401 -- not used for cryptographic purposes
```
[the remark this answers is not in the document]
Comment (jopit) [reply appstudio-controller/controllers/appstudio.redhat.com/promotionrun_controller.go]: Done.

**Review by jgwest — APPROVED** [review]
Looks great, thanks again @jopit!

**Pull request merged**